src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.cs

```diff
@@ -88,6 +89,17 @@ protected void Publish()
                 return;
             }
 
+            // MeterListener has a static constructor that creates runtime metrics instruments.
+            // We need to ensure this static constructor is called before starting to publish the instrument.
+            // This is necessary because creating runtime metrics instruments will cause re-entry to the Publish method,
+            // potentially resulting in a deadlock due to the SyncObject lock.
+            // Sequence of the deadlock:
+            //   1. An application creates an early instrument (e.g., Counter) before the MeterListener static constructor is executed.
+            //   2. Instrument.Publish is called and enters the SyncObject lock.
+            //   3. Within the lock block, MeterListener is called, triggering its static constructor.
+            //   4. The static constructor creates runtime metrics instruments, causing re-entry to Instrument.Publish and leading to a deadlock.
+            RuntimeHelpers.RunClassConstructor(typeof(MeterListener).TypeHandle);
```

Comment:
`RuntimeHelpers.RunClassConstructor` is reflection API. It works, but it is unnecessary overhead. It is not great that it will be called for every Publish.

A more efficient way to ensure static constructor trigger is a dummy Initialize method. We do that for example here: https://github.com/dotnet/runtime/blob/54e15cb81d84fae6760b90287468b09250f63e8c/src/libraries/Common/src/Interop/Unix/System.Net.Security.Native/Interop.Initialization.cs#L35-L38

Reply:
Is Publish hot? By the code it looked to me it is more like `register` and is called once for every kind of Meter, is it?

Reply:
Publish() is called once per instrument, likely 10-100 times in a run of typical web service process. I don't consider it hot.